File: frontend/src/types/topology.ts
// Frontend data model — kept consistent with the Go backend model

export interface Topology {
  project: Project;
  domains: Domain[];
  nodes: Node[];
  edges: Edge[];
  route_policies?: RoutePolicy[];
  // Allocation-scheme version number: written by the compiler and echoed back verbatim, so that a future
  // pin-format change can detect/migrate old allocations.
  // See docs/spec/compiler/allocation-stability.md (invariant I10).
  alloc_schema_version?: number;
}

export interface Project {
  id: string;
  name: string;
  description?: string;
  version?: string;
}

export interface Domain {
  id: string;
  name: string;
  cidr: string;
  description?: string;
  allocation_mode: 'auto' | 'manual';
  routing_mode: 'static' | 'babel' | 'none';
  reserved_ranges?: string[];
  // Per-domain transit address pool, defaulting to 10.10.0.0/24; transit IPs are now allocated per domain from this CIDR.
  // Corresponds to Domain.TransitCIDR in the Go model; see docs/spec/api/wire-contract.md (Domain field table).
  transit_cidr?: string;
}

export interface Node {
  id: string;
  name: string;
  hostname?: string;
  platform?: 'debian' | 'ubuntu';
  role: 'peer' | 'router' | 'relay' | 'gateway' | 'client';
  // Deployment mode (controller mode): 'managed' (default/empty) = agent-managed; 'manual' =
  // hand-deployed, no agent, carries its own pre-known public key + endpoint. Orthogonal to role.
  // See implementation_plans/mixed-controller-local-mode-2026_06_25.
  deployment_mode?: 'managed' | 'manual';
  domain_id: string;
  overlay_ip?: string;
  mtu?: number;
  // XDP attach mode for mimic (transport=tcp): empty/'skb' = generic XDP (default, best compatibility);
  // 'native' = driver-level XDP (faster, requires NIC support). See docs/spec/artifacts/mimic.md for details.
  xdp_mode?: 'skb' | 'native';
  // Overrides the auto-detected mimic egress interface ("" / omitted = auto-detect from the default
  // route). Set for multi-homed / policy-routing nodes (e.g. "wan0"). Only meaningful on a tcp link.
  mimic_egress_interface?: string;
  // Babel router-id; auto-generated from the node id when empty (see internal/renderer/babel.go).
  // MAC-48 form (e.g. 02:11:22:33:44:55) or an IPv4 address; meaningless for the client role.
  router_id?: string;
  capabilities: NodeCapabilities;
  fixed_private_key?: boolean;
  wireguard_private_key?: string;
  wireguard_public_key?: string;
  public_endpoints?: PublicEndpoint[];
  extra_prefixes?: string[];
  // SSH connection details (for auto-deploy)
  ssh_alias?: string;
  ssh_host?: string;
  ssh_port?: number;
  ssh_user?: string;
  ssh_key_path?: string;
  // Optional active-connectivity checks run by this managed node. Each signed policy names one
  // destination host and, for TCP, one port; it is deliberately not an arbitrary URL/command.
  telemetry_probes?: TelemetryProbe[];
}

export interface TelemetryProbe {
  id: string;
  type: 'icmp' | 'tcp';
  host: string;
  port?: number;
  interval_seconds?: number;
  timeout_milliseconds?: number;
}

export interface PublicEndpoint {
  id: string;
  host: string;
  port: number;
  note?: string;
}

export interface NodeCapabilities {
  can_accept_inbound: boolean;
  can_forward: boolean;
  can_relay: boolean;
  has_public_ip: boolean;
}

export interface Edge {
  id: string;
  from_node_id: string;
  to_node_id: string;
  type: 'direct' | 'public-endpoint' | 'relay-path' | 'candidate';
  endpoint_host?: string;
  endpoint_port?: number;  // user input: 0 = auto, nonzero = NAT override
  compiled_port?: number;  // read-only: actual port set by compiler
  priority?: number;
  weight?: number;
  // Link role for parallel links: 'backup' forms its own independent link (independent WG interface/port/transit/link-local address),
  // while no role or 'primary' belongs to the "primary-link class" (A->B + B->A for the same node pair are still merged into a single bidirectional tunnel).
  // See docs/spec/data-model/edge.md (§Parallel links).
  role?: 'primary' | 'backup';
  transport?: 'udp' | 'tcp';
  // Per-link mimic→UDP fallback POLICY (plan-4). undefined/omitted = inherit fleet default; 'udp' =
  // fall back to plain UDP if mimic provisioning fails; 'none' = fail closed. Only meaningful on a
  // tcp edge. See docs/spec/data-model/edge.md §TCP transport.
  mimic_fallback?: 'udp' | 'none';
  // Per-edge dial-direction POLICY. undefined/omitted/'both' = both sides may initiate (today's
  // behavior); 'forward' = only from→to initiates (dials the required endpoint_host), the reverse
  // peer keeps its [Peer] stanza but carries no Endpoint. There is no 'reverse' (D11, one
  // spelling): single-linking the other way is expressed by flipping the edge. Pure policy: gates
  // only which peer gets a dial Endpoint, never allocation. See docs/spec/data-model/edge.md
  // §Link direction.
  link_direction?: 'both' | 'forward';
  is_enabled: boolean;
  notes?: string;
  // Allocation pins: written by the compiler and echoed back verbatim, so that a recompile preserves existing
  // allocations (port / transit IP / link-local address), so adding new nodes does not disturb existing links. See
  // docs/spec/compiler/allocation-stability.md.
  pinned_from_port?: number;
  pinned_to_port?: number;
  pinned_from_transit_ip?: string;
  pinned_to_transit_ip?: string;
  pinned_from_link_local?: string;
  pinned_to_link_local?: string;
}

// RESERVED feature: route_policies is not wired into any renderer yet, and semantic validation rejects a non-empty array.
// This type is kept only for wire compatibility — do not build features on top of it.
// See docs/spec/api/wire-contract.md ("route_policies is RESERVED").
export interface RoutePolicy {
  id: string;
  domain_id: string;
  destination_cidr: string;
  next_hop_node_id?: string;
  metric?: number;
  notes?: string;
  source_selector?: string;
  action?: 'allow' | 'deny' | 'metric-override';
  apply_to_node_id?: string;
}

// API response types
export interface ValidationError {
  field: string;
  // code + params drive client localization via the 'error.<code>' catalog (tValidationError,
  // plan-3.5a); message is the server-rendered English default (CLI/curl + i18n fallback).
  code?: string;
  params?: Record<string, string>;
  message: string;
  level: 'error' | 'warning';
}

export interface ValidateResponse {
  valid: boolean;
  errors?: ValidationError[];
  warnings?: ValidationError[];
}

export interface CompileManifest {
  project_id: string;
  project_name: string;
  version: string;
  compiled_at: string;
  node_count: number;
  checksum: string;
}

export interface CompileResponse {
  topology: Topology;
  wireguard_configs: Record<string, string>;
  babel_configs: Record<string, string>;
  sysctl_configs: Record<string, string>;
  install_scripts: Record<string, string>;
  deploy_scripts: Record<string, string>;
  manifest: CompileManifest;
  // Non-fatal advisories (double NAT, edges missing an endpoint, orphan nodes, etc.).
  // The compile path runs semantic validation and returns these warnings alongside; the UI shows them after a successful compile.
  warnings?: ValidationError[];
  // Returned only by the controller compile preview (PR6): IDs of nodes that exist in the topology but were excluded
  // from this render because they are not yet enrolled.
  // The panel uses this to warn "N nodes not yet enrolled (not compiled)". The air-gap /api/compile does not return this field.
  skipped_unenrolled?: string[];
}

export interface CompileHistoryEntry {
  id: string;
  timestamp: string;
  topology: Topology;
  compileResult: CompileResponse;
}

File: internal/model/topology.go
package model

// Topology 完整的网络拓扑定义
type Topology struct {
	// 项目信息
	Project Project `json:"project"`

	// 网络域列表
	Domains []Domain `json:"domains"`

	// 节点列表
	Nodes []Node `json:"nodes"`

	// 边（连接）列表
	Edges []Edge `json:"edges"`

	// 路由策略（可选）
	RoutePolicies []RoutePolicy `json:"route_policies,omitempty"`

	// 分配方案版本号（粘性 pin 分配的 schema 版本，见 I10）。
	// 让未来对 pin 格式的改动能够检测并迁移旧拓扑，而不是把旧格式当成新格式静默误读。
	// 编译器在写回时统一标记为 compiler.AllocationSchemaVersion（当前为 1）。
	// 详见 docs/spec/compiler/allocation-stability.md（不变量 I10）。
	AllocSchemaVersion int `json:"alloc_schema_version,omitempty"`
}

// Project 项目元信息
type Project struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	Version     string `json:"version,omitempty"`
}

// Domain 网络域定义，代表一个 overlay 地址空间
type Domain struct {
	ID             string   `json:"id"`
	Name           string   `json:"name"`
	CIDR           string   `json:"cidr"`
	Description    string   `json:"description,omitempty"`
	AllocationMode string   `json:"allocation_mode"` // "auto" | "manual"
	RoutingMode    string   `json:"routing_mode"`    // "static" | "babel" | "none"
	ReservedRanges []string `json:"reserved_ranges,omitempty"`

	// Transit（点对点链路）地址池，用于 per-peer WireGuard 接口地址分配
	// 为空时自动使用 10.10.0.0/24
	TransitCIDR string `json:"transit_cidr,omitempty"`
}

// Node 网络节点定义
type Node struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Hostname string `json:"hostname,omitempty"`
	Platform string `json:"platform,omitempty"` // "debian" | "ubuntu"

	// 角色：peer, router, relay, gateway, client
	Role string `json:"role"`

	// 所属 Domain ID（必须引用已有 Domain）
	DomainID string `json:"domain_id"`

	// Overlay IP，编译时自动分配
	OverlayIP string `json:"overlay_ip,omitempty"`

	// WireGuard 接口 MTU（0 = 使用系统默认值，通常 1420）
	MTU int `json:"mtu,omitempty"`

	// mimic（transport=="tcp"）的 XDP 附着模式：空 / "skb" / "native"。
	// 空与 "skb" 等价：通用（generic）XDP，兼容几乎所有网卡（含不支持 native XDP 的
	// VPS virtio 网卡），是默认值。"native" 为驱动级 XDP，更快但需网卡/驱动支持；
	// 仅当操作员确认该节点网卡支持时才设置。详见 docs/spec/artifacts/mimic.md。
	XDPMode string `json:"xdp_mode,omitempty"`

	// Babel router-id（MAC-48 格式，如 02:11:22:33:44:55）
	// 留空时由编译器自动生成
	RouterID string `json:"router_id,omitempty"`

	// 节点能力
	Capabilities NodeCapabilities `json:"capabilities"`

	// 是否固定 WireGuard 密钥
	FixedPrivateKey bool `json:"fixed_private_key,omitempty"`

	// WireGuard 密钥（仅当 fixed_private_key=true 时有效）
	WireGuardPrivateKey string `json:"wireguard_private_key,omitempty"`
	WireGuardPublicKey  string `json:"wireguard_public_key,omitempty"`

	// 公网可达地址映射（一个节点可以有多组 endpoint）
	PublicEndpoints []PublicEndpoint `json:"public_endpoints,omitempty"`

	// 额外路由前缀（gateway 用：向外宣告的网段）
	ExtraPrefixes []string `json:"extra_prefixes,omitempty"`

	// SSH 连接信息（用于自动部署）
	SSHAlias   string `json:"ssh_alias,omitempty"`    // ssh_config 中的 Host 别名
	SSHHost    string `json:"ssh_host,omitempty"`     // SSH 主机地址
	SSHPort    int    `json:"ssh_port,omitempty"`     // SSH 端口（默认 22）
	SSHUser    string `json:"ssh_user,omitempty"`     // SSH 用户名
	SSHKeyPath string `json:"ssh_key_path,omitempty"` // SSH 私钥路径
}

// PublicEndpoint 公网可达地址映射
type PublicEndpoint struct {
	ID   string `json:"id"`
	Host string `json:"host"`
	Port int    `json:"port"`
	Note string `json:"note,omitempty"`
}

// NodeCapabilities 节点能力声明
type NodeCapabilities struct {
	CanAcceptInbound bool `json:"can_accept_inbound"`
	CanForward       bool `json:"can_forward"`
	CanRelay         bool `json:"can_relay"`
	HasPublicIP      bool `json:"has_public_ip"`
}

// Edge.Role 的取值常量。空值与 EdgeRolePrimary 等价（同归 primary class）。
const (
	// EdgeRolePrimary 主链路：与其它非 backup edge 一起折叠为一对节点的唯一一条链路。
	EdgeRolePrimary = "primary"
	// EdgeRoleBackup 备份链路：每条 backup edge 各自成为一条独立链路（独立分配 + 独立接口名）。
	EdgeRoleBackup = "backup"
)

// Edge 边定义，表示两个节点之间的连接意图
// 语义为 "from 主动连 to "
type Edge struct {
	ID         string `json:"id"`
	FromNodeID string `json:"from_node_id"`
	ToNodeID   string `json:"to_node_id"`

	// 连接类型：direct, public-endpoint, relay-path, candidate
	Type string `json:"type"`

	// 对端 endpoint（用户输入：0 = 自动分配，非零 = NAT/端口转发覆盖）
	EndpointHost string `json:"endpoint_host,omitempty"`
	EndpointPort int    `json:"endpoint_port,omitempty"`

	// 编译器分配的实际端口（只读，由编译器填充）
	CompiledPort int `json:"compiled_port,omitempty"`

	// 优先级
	Priority int `json:"priority,omitempty"`
	Weight   int `json:"weight,omitempty"`

	// 链路角色：区分一对节点之间的主链路（primary class）与备份链路。
	// 空值 / "primary" 归入 primary class——同一对节点的所有非 backup edge 折叠为一条链路；
	// "backup" 则每条 edge 各自成为一条独立链路，用于 Babel 基于 cost 的故障切换。
	// 详见 docs/spec/compiler/allocation-stability.md（Link identity with parallel edges）。
	Role string `json:"role,omitempty"`

	// 传输协议：udp, tcp
	Transport string `json:"transport,omitempty"`

	// 是否启用
	IsEnabled bool `json:"is_enabled"`

	// 备注
	Notes string `json:"notes,omitempty"`

	// === 分配 pin（读写）：编译器写回、前端持久化并原样回传 ===
	// 这六个字段把本链路分配到的资源（监听端口、transit IP 对、IPv6 link-local 对）
	// 绑定到具体的 edge 上，复用 overlay_ip / compiled_port 已有的「写回 + localStorage」
	// 往返路径（无新增传输）。下次编译时编译器优先沿用已存在的 pin，而非按数组位置重算，
	// 从而让 superset 拓扑对每条既有 edge 重现逐字节相同的分配值（增量扩展不打扰未变动的节点）。
	//
	// pin 按所属 edge 的 FromNodeID/ToNodeID 定向：from_* 是本 edge from 侧的值，
	// to_* 是 to 侧的值；同一对节点的反向 edge 携带的是镜像后的同一对取值。
	// 每个资源要么成对完整 pin、要么完全不 pin；单端 pin（partial）由验证器拒绝。
	// 详见 docs/spec/compiler/allocation-stability.md 与 docs/spec/data-model/edge.md（Allocation pins）。

	// from / to 侧接口的监听端口 pin（client 角色不使用 per-peer 端口，故不 pin）。
	PinnedFromPort int `json:"pinned_from_port,omitempty"`
	PinnedToPort   int `json:"pinned_to_port,omitempty"`

	// from / to 侧的 transit IP pin（取自所属域的 transit_cidr 地址池）。
	PinnedFromTransitIP string `json:"pinned_from_transit_ip,omitempty"`
	PinnedToTransitIP   string `json:"pinned_to_transit_ip,omitempty"`

	// from / to 侧的 IPv6 link-local pin（取自 fe80::/10）。
	PinnedFromLinkLocal string `json:"pinned_from_link_local,omitempty"`
	PinnedToLinkLocal   string `json:"pinned_to_link_local,omitempty"`
}

// RoutePolicy 路由策略定义
type RoutePolicy struct {
	ID              string `json:"id"`
	DomainID        string `json:"domain_id"`
	DestinationCIDR string `json:"destination_cidr"`
	NextHopNodeID   string `json:"next_hop_node_id,omitempty"`
	Metric          int    `json:"metric,omitempty"`
	Notes           string `json:"notes,omitempty"`

	SourceSelector string `json:"source_selector,omitempty"`
	Action         string `json:"action,omitempty"`
	ApplyToNodeID  string `json:"apply_to_node_id,omitempty"`
}
